torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -707,7 +707,8 @@ def reset_sharded_param(self):
                     f"Expects swap_tensors to preserve object but got {new_param} "
                     f"instead of {self.sharded_param}"
                 )
-            self.sharded_param = new_param
+        elif skip_same_param:
+            return
```

Comment:
reset_sharded_param has 3 callsites
* state dict load
* _apply, eg model.to(dtype)
* lazy_init

for lazy_init, we should return early if the param stays the same. otherwise at the end we create new param with `nn.Parameter()` and that will obsolete optim's reference to `model.parameters()`

Reply:
Wait, so is `main` branch wrong today ignoring this PR?

FSDP's `self.sharded_param` is a different tensor than the one referenced by the optimizer?

If so, I think we should fix this bug separately.

Reply:
haha. trunk is healthy because we do `self.sharded_param._local_tensor = local_tensor[: self.sharded_size[0]]` in trunk. that won't create a new `nn.Parameter()`